### PR TITLE
[SPARK-35135][CORE] Turn the `WritablePartitionedIterator` from a trait into a default implementation class 

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
`WritablePartitionedIterator` define in `WritablePartitionedPairCollection.scala` and there are two implementation of these trait,  but the code for these two implementations is duplicate.

The main change of this pr is turn the `WritablePartitionedIterator` from a trait into a default implementation class because there is only one implementation now.

### Why are the changes needed?
Cleanup duplicate code.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass the Jenkins or GitHub Action

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala [hunk 1]
```diff
@@ -94,3 +83,20 @@ private[spark] trait WritablePartitionedIterator {
 
   def nextPartition(): Int
 }
+
+/**
+ * Default implementation of `WritablePartitionedIterator`.
+ */
+private[spark] class DefaultWritablePartitionedIterator[K, V](it: Iterator[((Int, K), V)])
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala:90]: Is there any suggestion for a better class name?

**Comment by HyukjinKwon** [comment]
cc @Ngone51 FYI

**Comment by LuciferYang** [comment]
@Ngone51 Do you have time to review this pr ? thx ~

File: core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala:90]: I think we can just turn the trait into the class if there's only one implementation now.

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala:90]: ok ~

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala]: turn the trait into the class will fail MiMa tests:
```
[error] spark-core: Failed binary compatibility check against org.apache.spark:spark-core_2.12:3.0.0! Found 1 potential problems (filtered 1184)
[error]  * declaration of interface org.apache.spark.util.collection.WritablePartitionedIterator is class org.apache.spark.util.collection.WritablePartitionedIterator in current version; changing interface to class breaks client code
[error]    filter with: ProblemFilters.exclude[IncompatibleTemplateDefProblem]("org.apache.spark.util.collection.WritablePartitionedIterator")
```

Can you give me some guidance on how to solve this problem? @Ngone51  @HyukjinKwon

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/util/collection/WritablePartitionedPairCollection.scala]: 8835ecc add a `IncompatibleTemplateDefProblem` exclude filter to `MimaExcludes.scala`, GA passed now. @Ngone51 @HyukjinKwon

**Review by HyukjinKwon — APPROVED** [review]

**Review by Ngone51 — APPROVED** [review]

**Comment by LuciferYang** [comment]
thx ~ @HyukjinKwon @Ngone51

**Comment by Ngone51** [comment]
Thanks, merged to Master!

**Pull request closed without merging**